pkg/apis/kubeflow.org/v2alpha1/trainjob_types.go

```diff
@@ -15,3 +15,229 @@ limitations under the License.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired TrainJob.
+	Spec TrainJobSpec `json:"spec,omitempty"`
+
+	// Current status of TrainJob.
+	Status TrainJobStatus `json:"status,omitempty"`
+}
+
+// TrainJobSpec represents specification of the desired TrainJob.
+type TrainJobSpec struct {
+	// Reference to the training runtime.
+	TrainingRuntimeRef TrainingRuntimeRef `json:"trainingRuntimeRef"`
+
+	// Configuration of the desired trainer.
+	Trainer *Trainer `json:"trainer,omitempty"`
+
+	// Configuration of the training dataset.
+	DatasetConfig *DatasetConfig `json:"datasetConfig,omitempty"`
+
+	// Configuration of the pre-trained and trained model.
+	ModelConfig *ModelConfig `json:"modelConfig,omitempty"`
+
+	// Labels to apply for the derivative JobSet and Jobs.
+	// They will be merged with the TrainingRuntime values.
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// Annotations to apply for the derivative JobSet and Jobs.
+	// They will be merged with the TrainingRuntime values.
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// Custom overrides for the training runtime.
+	PodSpecOverrides []PodSpecOverrides `json:"podSpecOverrides,omitempty"`
+
+	// Whether the controller should suspend the running TrainJob.
+	// Defaults to false.
+	Suspend *bool `json:"suspend,omitempty"`
+
+	// ManagedBy field indicates the controller that manages a TrainJob.
+	ManagedBy *string `json:"managedBy,omitempty"`
+}
+
+// TrainingRuntimeRef represents the reference to the existing training runtime.
+type TrainingRuntimeRef struct {
+	// Name of the runtime being referenced.
+	// When namespaced-scoped TrainingRuntime is used, the TrainJob must have
+	// the same namespace as the deployed runtime.
+	Name string `json:"name"`
+
+	// APIGroup of the runtime being referenced.
+	// Defaults to `kubeflow.org`.
+	APIGroup *string `json:"apiGroup,omitempty"`
+
+	// Kind of the runtime being referenced.
+	// It must be one of TrainingRuntime or ClusterTrainingRuntime.
+	// Defaults to ClusterTrainingRuntime.
+	Kind *string `json:"kind,omitempty"`
+}
+
+// Trainer represents the desired trainer configuration.
+// Every training runtime contains `trainer` container which represents Trainer.
+type Trainer struct {
+	// Docker image for the training container.
+	Image string `json:"image,omitempty"`
+
+	// Entrypoint commands for the training container.
+	Command []string `json:"command,omitempty"`
+
+	// Arguments to the entrypoint for the training container.
+	Args []string `json:"args,omitempty"`
+
+	// List of environment variables to set in the training container.
+	// These values will be merged with the TrainingRuntime's trainer environments.
+	Env []corev1.EnvVar `json:"env,omitempty"`
+
+	// Number of training nodes.
+	// TODO (andreyvelich): Do we want to support dynamic num of nodes in TrainJob for PyTorch elastic: `--nnodes=1:4` ?
+	NumNodes *int32 `json:"numNodes,omitempty"`
```

Comment:
@tenzen-y What do you think about using `*string` here since for the elastic training user can set:
```
torchrun --nnodes=1:4
```

Reply:
IIRC, we can specify the elastically policy in the JobSetSpec. So, let's use the typed int here.
After we find some advantages for string, we can introduce IntOrString like Deployment rollingUpdate.